File: Python/36_Valid_Sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row,col=9,9

        def isValidRow(r):
            res=set()
            for c in range(9):
                if board[r][c] != '.' and board[r][c] in res:
                    return False
                res.add(board[r][c])
            return True
        def isValidCol(c):
            res=set()
            for r in range(9):
                if board[r][c] !='.' and board[r][c] in res:
                    return False
                res.add(board[r][c])
            return True
        def isValidBox(r,c):
            res=set()
            for i in range(r,r+3):
                for j in range(c,c+3):
                    if board[i][j] !='.' and board[i][j] in res:
                        return False
                    res.add(board[i][j])
            return True
        
        for i in range(9):
            if isValidRow(i)==False:
                return False
            if isValidCol(i)==False:
                return False
        for i in range(3):
            for j in range(3):
                if isValidBox(i*3,j*3)==False:
                    return False
        return True

File: Java/36_Valid_Sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        for(int i=0;i<9;i++){
            if(isValidRow(board,i)==false){
                return false;
            }
            if(isValidColumn(board,i)==false){
                return false;
            }
        }
        for(int i=0;i<3;i++){
            for(int j=0;j<3;j++){
                if(isValidCube(board,i*3,j*3)==false){
                    return false;
                }
            }
        }
        return true;
    }
    private boolean isValidColumn(char[][] board,int column){
        HashSet<Character> nums=new HashSet<>();
        for(int r=0;r<9;r++){
            if(nums.contains(board[r][column])&& board[r][column]!='.'){
                return false;
            }
            nums.add(board[r][column]);
        }
        return true;
    }
    private boolean isValidRow(char[][]board,int row){
        HashSet<Character> nums=new HashSet<>();
        for(int c=0;c<9;c++){
            if(nums.contains(board[row][c])&& board[row][c]!='.'){
                return false;
            }
            nums.add(board[row][c]);
        }
        return true;
    }
    private boolean isValidCube(char[][]board, int row,int column){
        HashSet<Character> nums=new HashSet<>();
        for(int r=0;r<3;r++){
            for(int c=0;c<3;c++){
                if(nums.contains(board[row+r][column+c])&& board[row+r][column+c]!='.'){
                    return false;
                }
                nums.add(board[row+r][column+c]);
            }
        }
        return true;
    }

}
